ouroboros-consensus/src/Ouroboros/Storage/ChainDB/VolDB.hs

```diff
@@ -0,0 +1,265 @@
+    -- add ticket for tracking the size of our in-memory structures.
+
+    -- 'hashes' constructs a list of hashes. We need to get ChainFragments of
+    -- something that impleemnts HasHeader. Getting there currently requires
+    -- reading blocks. We might need to reconsider this.
+    mkFragments :: [HeaderHash blk] -> m (ChainFragment hdr)
+    mkFragments = fmap (Fragment.fromOldestFirst) . mapM (getKnownHeader db)
+
+    -- Construct fragments of hashes
+    --
+    -- We do this as a first step since this is pure function
+    hashes :: ChainHash blk
+           -> (Maybe (HeaderHash blk) -> Set (HeaderHash blk))
+           -> [[HeaderHash blk]]
+    hashes prev succsOf = do
+        next <- Set.toList $ succsOf (fromChainHash prev)
+        rest <- hashes (BlockHash next) succsOf
+        return $ next : rest
+
+{-------------------------------------------------------------------------------
+  Getting and parsing blocks
+-------------------------------------------------------------------------------}
+
+getKnownHeader :: (MonadCatch m, StandardHash blk, Typeable blk)
+               => VolDB m blk hdr -> HeaderHash blk -> m hdr
+getKnownHeader db@VolDB{..} hash =
+    header <$> getKnownBlock db hash
+
+getKnownBlock :: (MonadCatch m, StandardHash blk, Typeable blk)
+              => VolDB m blk hdr -> HeaderHash blk -> m blk
+getKnownBlock db hash = do
+    mBlock <- mustExist hash <$> getBlock db hash
+    case mBlock of
+      Right b  -> return b
+      Left err -> throwM err
+
+getBlock :: (MonadCatch m, StandardHash blk, Typeable blk)
+         => VolDB m blk hdr -> HeaderHash blk -> m (Maybe blk)
+getBlock db hash = do
+    mBlob <- withDB db $ \vol ->
+               fmap (parse (decBlock db) hash) <$> VolDB.getBlock vol hash
+    case mBlob of
+      Nothing         -> return $ Nothing
+      Just (Right b)  -> return $ Just b
+      Just (Left err) -> throwM $ err
+
+{-------------------------------------------------------------------------------
+  Auxiliary: parsing
+-------------------------------------------------------------------------------}
+
+blockFileParser :: forall m blk hdr. (MonadST m, MonadThrow m, HasHeader blk)
+                => VolDbArgs m blk hdr
+                -> VolDB.Parser
+                     (VolDB.ParserError (HeaderHash blk))
+                     m
+                     (HeaderHash blk)
+blockFileParser VolDbArgs{..} =
+    VolDB.Parser $
+        fmap (second (fmap mkErr))
+      . Util.CBOR.readIncrementalOffsets volHasFS decoder'
+  where
+    decoder' :: forall s. Decoder s (VolDB.BlockInfo (HeaderHash blk))
+    decoder' = extractInfo <$> volDecodeBlock
+
+    extractInfo :: blk -> VolDB.BlockInfo (HeaderHash blk)
+    extractInfo b = VolDB.BlockInfo {
+          bbid    = blockHash b
+        , bslot   = blockSlot b
+        , bpreBid = fromChainHash (blockPrevHash b)
+        }
+
+    -- TODO: This won't be needed anymore after Kostas' next PR
```

Comment:
Actually I think Kostas' PR doesn't yet cover this.

Reply:
(There is a workaround, one of the `ParserError` constructors takes ` String`, so can `show` and then use that as a stop-gap measure.)

Reply:
#488.